feat: allow filtering of validators by top-level validator status

vladmarusyk opened the pull request:
## Add support for filtering validators by top-level status

This pull request implements support for filtering validators by top-level statuses such as `active`, `pending`, `exited`, and `withdrawal` as specified in the beacon-api spec.

### Changes:
- Added `mapToGeneralStatus` function to map detailed statuses to top-level statuses (`active`, `pending`, `exited`, and `withdrawal`).
- Updated `filterStateValidatorsByStatus` to support filtering validators by both detailed and top-level statuses.

### Related issue:
Closes #7141

vladmarusyk committed:
Fix ValidatorStatus type issue

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ChainSafe/lodestar?pullRequest=7143) <br/>All committers have signed the CLA.

nflaig commented on packages/beacon-node/src/api/impl/beacon/state/utils.ts:
```diff
@@ -83,6 +83,51 @@ export async function getStateResponseWithRegen(
   return res;
 }
 
+function mapToGeneralStatus(subStatus: string): string {
```
switch-case would be nicer here, we can also reuse `ValidatorStatus` type, and potentially a extended type which includes top-level / general statuses as well

nflaig commented on packages/beacon-node/src/api/impl/beacon/state/utils.ts:
```diff
@@ -83,6 +83,51 @@ export async function getStateResponseWithRegen(
+ * Get the status of the validator
+ * based on conditions outlined in https://hackmd.io/ofFJ5gOmQpu1jjHilHbdQQ
+ */
+export function getValidatorStatus(validator: phase0.Validator, currentEpoch: Epoch): routes.beacon.ValidatorStatus {
```
why is this added here?

nflaig reviewed (commented):
Thanks for picking this up, some initial feedback

nflaig commented:
also note that `ValidatorStatus` type was just moved to types packages (see https://github.com/ChainSafe/lodestar/pull/7140), make sure your branch is up-to-date

vladmarusyk committed:
refactor: use switch-case and typed general validator status mapping

vladmarusyk commented:
I've updated the code based on the feedback:
- Replaced `if-else` with `switch-case`.
- Used the `ValidatorStatus` type and created an extended type for general statuses.

nflaig commented on packages/beacon-node/src/api/impl/beacon/state/utils.ts:
```diff
@@ -83,6 +83,60 @@ export async function getStateResponseWithRegen(
   return res;
 }
 
+export type GeneralValidatorStatus = routes.beacon.ValidatorStatus | "active" | "pending" | "exited" | "withdrawal";
```
On 2nd thought it makes more sense to just declare the top-level statuses here

```suggestion
export type GeneralValidatorStatus = "active" | "pending" | "exited" | "withdrawal";
```

nflaig reviewed (commented):
Looks good, can you please remove unrelated code updates, see https://github.com/ChainSafe/lodestar/pull/7143#discussion_r1794055386

vladmarusyk committed:
Update packages/beacon-node/src/api/impl/beacon/state/utils.ts

Co-authored-by: Nico Flaig <nflaig@protonmail.com>

vladmarusyk committed:
refactor: deleted unrelated code

nflaig commented:
https://github.com/ChainSafe/lodestar/pull/7143/commits/e07a82be26ab5c1a3258a5939dcac67d72b1be33 just removed the actual code, the unrelated code changes are still here

vladmarusyk commented:
Sorry for the inconvenience, solving the first issue. Thank you for the opportunity.

vladmarusyk committed:
refactor: deleted unrelated code

vladmarusyk commented:
Removed unrelated code

nflaig commented on packages/beacon-node/src/api/impl/beacon/state/utils.ts:
```diff
@@ -83,6 +83,28 @@ export async function getStateResponseWithRegen(
   return res;
 }
 
+export type GeneralValidatorStatus = "active" | "pending" | "exited" | "withdrawal";
```
no need to export this type

```suggestion
type GeneralValidatorStatus = "active" | "pending" | "exited" | "withdrawal";
```

nflaig reviewed (commented):
LGTM - thanks for the cleanup

vladmarusyk committed:
Update packages/beacon-node/src/api/impl/beacon/state/utils.ts

Co-authored-by: Nico Flaig <nflaig@protonmail.com>

wemeetagain reviewed (approved):
LGTM

unrelated, commenting for posterity
 `ValidatorStatus` shouldn't contain `"active"`

nflaig commented:
> unrelated, commenting for posterity `ValidatorStatus` shouldn't contain `"active"`

yeah, let's get rid of this https://github.com/ChainSafe/lodestar/pull/7146

wemeetagain commented:
:tada: This PR is included in [v1.23.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.23.0) :tada:
